numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -2986,9 +2986,13 @@ array_einsum(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
     }
 
     /* Get the keyword arguments */
+    int error = 0;
     if (kwds != NULL) {
         PyObject *key, *value;
         Py_ssize_t pos = 0;
+#if Py_GIL_DISABLED
+        Py_BEGIN_CRITICAL_SECTION(kwds);
+#endif
```

Comment:
This is keyword parsing, you don't need it:
> keyword argument dictionaries in function calls; those keyword argument dictionaries are effectively private (not accessible by other threads), so using borrowed references in that context is safe.

Reply:
(Although we should just use `npy_parse_kwargs` or whatever it was here...  Will make this much faster and cleaner.)

Reply:
Ah good point, thank you for helping me spot these forest-for-the-trees issues.